hazelcast/include/hazelcast/client/SocketInterceptor.h

```diff
@@ -15,25 +15,57 @@
  */
 #pragma once
 
+#include <functional>
+#include <utility>
+
 #include "hazelcast/util/HazelcastDll.h"
+#include "hazelcast/util/noop.h"
+#include "hazelcast/util/type_traits.h"
 
 namespace hazelcast {
     namespace client {
         class Socket;
 
+        namespace connection {
+            class ClientConnectionManagerImpl;
+        }
+
         /**
-         * Base class for socketInterceptor classes to inherit from.
+         * An interface that provides the ability to intercept the creation of sockets.
          *
-         * @see ClientConfig#setSocketInterceptor(SocketInterceptor *socketInterceptor);
+         * \see ClientConfig::setSocketInterceptor
          */
-        class HAZELCAST_API SocketInterceptor {
+        class HAZELCAST_API SocketInterceptor final {
         public:
             /**
-             * Will be called with the Socket, each time client creates a connection to any Member.
+             * Set an handler function that will be called with a Socket, 
+             * each time the client creates a connection to any Member.
+             * \param h a `void` function object that is callable with a single parameter of type `const Socket &`
+             */
+            template<typename Handler,
+                     typename = util::enable_if_rvalue_ref_t<Handler &&>>
+            SocketInterceptor &on_connect(Handler &&h) & {
+                connect = std::forward<Handler>(h);
+                return *this;
+            }
+
+            /**
+             * \copydoc SocketInterceptor::on_connect
              */
-            virtual void onConnect(const Socket &connectedSocket) = 0;
+            template<typename Handler,
+                     typename = util::enable_if_rvalue_ref_t<Handler &&>>
+            SocketInterceptor &&on_connect(Handler &&h) && {
+                on_connect(std::forward<Handler>(h));
+                return std::move(*this);
+            }
+
+
+        private:
+            friend class connection::ClientConnectionManagerImpl;
+
+            using handler_t = std::function<void(const Socket &)>;
 
-            virtual ~SocketInterceptor();
+            handler_t connect{ util::noop<const Socket &> };
```

Comment:
private member naming `connect` -> `connect_`. Seems like we also forget this for other listeners as well. It is not very critical.

Reply:
Ignore that for now, I've been able to automatically refactor all private data member names to in `lower_case_` format using clang-tidy: https://github.com/hazelcast/hazelcast-cpp-client/pull/632. They'll be fixed once we merge that.